### PR TITLE
Improve C++20 detection with CMake 3.12+ and gcc-9/c++20 support fixes

### LINKED ISSUE #859: Build fails with -DWITH_STL=ON cmake option
**Describe your environment** 
gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)

**Steps to reproduce**
```bash
~/ot/build$cmake -DWITH_STL=ON -DCMAKE_CXX_STANDARD=20 ..
~/ot/build$ make -j${nproj}
```
**What is the expected behavior?**
Build should be successful

**What is the actual behavior?**
Build fails with below error:

```
[  1%] Built target timestamp_test
[  2%] Built target runtime_context_test
[  3%] Built target context_test
[  5%] Built target composite_propagator_test
[  7%] Building CXX object api/test/nostd/CMakeFiles/variant_test.dir/variant_test.cc.o
[  7%] Built target dynamic_load_test
[  8%] Built target function_ref_test
[  9%] Building CXX object api/test/nostd/CMakeFiles/string_view_test.dir/string_view_test.cc.o
Scanning dependencies of target unique_ptr_test
Scanning dependencies of target utility_test
[  9%] Building CXX object api/test/nostd/CMakeFiles/unique_ptr_test.dir/unique_ptr_test.cc.o
[ 10%] Building CXX object api/test/nostd/CMakeFiles/utility_test.dir/utility_test.cc.o
In file included from /usr/include/gtest/gtest.h:62,
                 from /home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:9:
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc: In member function ‘virtual void VariantTest_Get_Test::TestBody()’:
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:52:45: error: ‘bad_variant_access’ in namespace ‘nostd’ does not name a type
   52 |   EXPECT_THROW(nostd::get<float>(w), nostd::bad_variant_access);
      |                                             ^~~~~~~~~~~~~~~~~~
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:52:3: error: expected ‘)’ before ‘const’
   52 |   EXPECT_THROW(nostd::get<float>(w), nostd::bad_variant_access);
      |   ^~~~~~~~~~~~
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:52:3: error: expected ‘{’ before ‘const’
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:52:3: error: expected unqualified-id before ‘)’ token
   52 |   EXPECT_THROW(nostd::get<float>(w), nostd::bad_variant_access);
      |   ^~~~~~~~~~~~
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/variant_test.cc:52:3: error: expected primary-expression before ‘catch’
   52 |   EXPECT_THROW(nostd::get<float>(w), nostd::bad_variant_access);
      |   ^~~~~~~~~~~~
In file included from /usr/include/gtest/gtest.h:375,
                 from /home/labhas/ot/opentelemetry-cpp/api/test/nostd/string_view_test.cc:6:
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/string_view_test.cc: In member function ‘virtual void StringViewTest_CStringInitialization_Test::TestBody()’:
/home/labhas/ot/opentelemetry-cpp/api/test/nostd/string_view_test.cc:26:32: error: ‘strlen’ is not a member of ‘std’; did you mean ‘strlen’?
   26 |   EXPECT_EQ(ref.length(), std::strlen(val));
      |                                ^~~~~~
In file included from /usr/include/gtest/internal/gtest-port.h:250,
                 from /usr/include/gtest/internal/gtest-internal.h:40,
                 from /usr/include/gtest/gtest.h:62,
                 from /home/labhas/ot/opentelemetry-cpp/api/test/nostd/string_view_test.cc:6:
/usr/include/string.h:385:15: note: ‘strlen’ declared here
  385 | extern size_t strlen (const char *__s)
      |               ^~~~~~
make[2]: *** [api/test/nostd/CMakeFiles/string_view_test.dir/build.make:63: api/test/nostd/CMakeFiles/string_view_test.dir/string_view_test.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1221: api/test/nostd/CMakeFiles/string_view_test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 10%] Linking CXX executable utility_test
make[2]: *** [api/test/nostd/CMakeFiles/variant_test.dir/build.make:63: api/test/nostd/CMakeFiles/variant_test.dir/variant_test.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1167: api/test/nostd/CMakeFiles/variant_test.dir/all] Error 2
[ 10%] Built target utility_test
[ 11%] Linking CXX executable unique_ptr_test
[ 11%] Built target unique_ptr_test
make: *** [Makefile:141: all] Error 2
```
**Additional context**

cc @maxgolov

### PR DESCRIPTION
**Pull request by maxgolov**
Fixes #859 

- CMake v3.12 introduced C++20 support :

image

Our logic was assuming that it's only introduced in 3.18.x , and it would break in Ubuntu-20.04 which has a compiler with C++20, but its CMake version is only 3.16 (less than 3.18).. Thus, it was always falling back to C++17 language standard.

- now when we actually force C++20 standard, there were two minor compilation issues. One in our `nostd::variant` projection - it was missing `bad_variant_access`. And the second one was using `std::strlen` without `#include <cstring>`. Both have now been resolved. Interestingly enough, those issues did not affect vs2019 in c++20/c++latest 😄 

**How to test**

Our build system "developer-focused" part for Linux and Mac (`/tools/`) already has the build scripts to run the build as follows:

**nostd:**
```console
tools/build.sh
```

**stdlib:**
```console
tools/build.sh stdlib -DWITH_STL=ON
```

That would produce the binaries under `out/x86_64-linux-gnu-gcc-9/[nostd|stdlib]`. Default build without any parameters - produces `nostd` flavor. And if custom parameters passed to CMake - these allow to produce the build with any configuration name (1st argument is configuration name). For example, the build with Standard Library types, which in turn always turns on C++20 , if available. If C++20 is not available, then it falls back to C++17.

Build script `tools/build.sh` - allows to build different configurations side-by-side in one workspace, with build artifacts for different configurations produced side-by-side in a single output directory, as follows:

image

It's also possible to use WSL2 and Windows machine to host multiple docker images with different compilers. All producing the output for subsequent packaging from one single binary artifacts directory a.k.a. multiple docker / single filesystem build farm, side-by-side, one workspace - many containers - separate output directories for artifacts built by different compilers. Neat, isn't it?

### CODE REVIEW

**Review by lalitb — APPROVED** [review]

**Comment by maxgolov** [comment]
@ThomsonTan - if you can have a quick look? I think this PR would be valuable addition to Lalit's CI PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -107,7 +107,7 @@ if(WITH_STL)
   add_definitions(-DHAVE_CPP_STDLIB)
   add_definitions(-DHAVE_GSL)
   # Require at least C++17. C++20 is needed to avoid gsl::span
-  if(CMAKE_VERSION VERSION_GREATER 3.18.0)
+  if(CMAKE_VERSION VERSION_GREATER 3.11.999)
```
Comment (ThomsonTan) [inline CMakeLists.txt:110]: Use `VERSION_GREATER_EQUAL 3.12.0` to avoid using the crafted version number?

**Review by ThomsonTan — APPROVED** [review]
Thanks for fixing this.

**Comment by lalitb** [comment]
> @ThomsonTan - if you can have a quick look? I think this PR would be valuable addition to Lalit's CI PR.

@ThomsonTan  - also I have addressed your review comments in #855. Please see if they are fine. Thanks

**Pull request merged**